[ALLUXIO-2416] Merge UFS development branch into master

madanadit opened the pull request:
Refactoring of the UFS API.

Summary of changes:
- exists is split into isFile and isDirectory
- delete, rename is split into deleteFile/deleteDirectory and renameFile/renameDirectory
- Options idiom is used consistently across the API
- Object store implementations have been abstracted
- list has been changed to listStatus

JIRA: https://alluxio.atlassian.net/browse/ALLUXIO-2416

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12463/
Test PASSed.

calvinjia commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -0,0 +1,138 @@
+   * @param uri the {@link AlluxioURI} used to create this ufs
+   */
+  protected BaseUnderFileSystem(AlluxioURI uri) {
+    Preconditions.checkNotNull(uri);
```
I think you can do `mUri = Preconditions.checkNotNull(uri)`

calvinjia commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -0,0 +1,138 @@
+    }
+    // Clean the path by creating a URI and turning it back to a string
+    AlluxioURI uri = new AlluxioURI(path);
+    path = uri.toString();
```
We should have a separate `validatePath` method for this

calvinjia commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -0,0 +1,138 @@
+    } else {
+      for (UnderFileStatus subp : subpaths) {
+        pathsToProcess.add(
+            new UnderFileStatus(PathUtils.concatPath(path, subp.getName()), subp.isDirectory()));
```
This seems like an abuse of `UnderFileStatus`, since, UnderFileStatus#getName returns the name of the resource, but we overload it with the full path here.

calvinjia commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -0,0 +1,138 @@
+      }
+    }
+    while (!pathsToProcess.isEmpty()) {
+      UnderFileStatus p = pathsToProcess.remove();
```
Preference for full variable names

calvinjia commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -0,0 +1,138 @@
+    AlluxioURI uri = new AlluxioURI(path);
+    path = uri.toString();
+    List<UnderFileStatus> returnPaths = new ArrayList<>();
+    Queue<UnderFileStatus> pathsToProcess = new ArrayDeque<>();
```
Any reason why we use a `Deque` when we use it as a `Queue`?

madanadit replied:
I did not right this code, but I think the double ended-ness comes at no extra cost. A LinkedList is an alternative but ArrayDeque seems to be preferred because of line caching.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** Maximum length for a single listing query. */
+  private static final int LISTING_LENGTH_MAX = 1000;
```
Should this be overridable by the subclasses instead of `LISTING_LENGTH`?

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+   */
+  protected class ObjectStatus {
+    long mFileSize;
+    long mLastModifiedTimeMs;
```
final?

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+   * Information about a single object in object UFS.
+   */
+  protected class ObjectStatus {
+    long mFileSize;
```
final? I think this should be called `mContentLength` just for consistency with the getter.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+      return createObject(stripPrefixIfPresent(path));
+    }
+    return null;
+  }
```
Should an exception be thrown instead of returning null?

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+    if (!options.isRecursive()) {
+      UnderFileStatus[] children = listInternal(path, ListOptions.defaults());
+      if (children == null) {
+        LOG.error("Unable to delete {} because listInternal returns null", path);
```
Maybe a better error message?

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+   * @param key ufs key to get metadata for
+   * @return {@link ObjectStatus} if key exists and successful, otherwise null
+   */
+  protected abstract ObjectStatus getObjectStatus(String key);
```
Should this only be called for files? or is it expected for the implementation to check both the key and the folder version of the key.

madanadit replied:
No, it can be called for a file or a directory. It is expected that the caller will change the key to a directory key if desired. This is in agreement with the general design that the implementations extending ObjectUnderFS do not need to worry about directory encoding (by default).

calvinjia replied:
Should `getModificationTimeMs` be modified to use the folder path if the file path does not exist?

madanadit replied:
Seems like the method is only used for files. I don't know if we want to have a method for directories. If we do, I suggest we split it up into a different function instead of overloading this one.

gpang commented on core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java:
```diff
@@ -191,24 +189,10 @@ private InputStream openAtPosition(long position) throws IOException {
-        }
+        UnderFileSystem ufs = UnderFileSystem.Factory.get(mUri);
+        mStream =
+            new CountingInputStream(ufs.open(mUri, OpenOptions.defaults().setOffset(position)));
```
This guarantees that the stream will be opened from the correct position?

Does that mean the block of code below is obsolete? @calvinjia

calvinjia replied:
The block of code below should still be used in the case `mStream` was already constructed and the position requested is ahead of the current position of the stream.

gpang commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+     *
+     * @return the file size in bytes
+     */
+    public long getContentLength() {
```
Why is this called content length, and not file length?

madanadit replied:
Well it is consistent with the terminology used by object stores as they do not refer to 'objects' as 'files'. I renamed the variable to be consistent. I can change it if you feel strongly about it.

gpang commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,659 @@
+    /**
+     * Objects in a pseudo-directory which may be a file or a directory.
+     *
+     * @return a list of objects
```
NIT: objects -> object names

madanadit commented:
@gpang @calvinjia Thanks for the review!. Its ready for another pass.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12472/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12472/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/12472/org.alluxio$alluxio-tests/testReport/alluxio.underfs/UnderStorageSystemInterfaceIntegrationTest/listStatusRecursive/'><strong>alluxio.underfs.UnderStorageSystemInterfaceIntegrationTest.listStatusRecursive</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12473/
Test PASSed.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+    final long mLastModifiedTimeMs;
+
+    public ObjectStatus(long fileSize, long lastModifiedTimeMs) {
+      mContentLength = fileSize;
```
Could you make the names consistent?

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+    }
+
+    /**
+     * Gets the UTC last modified time in ms.
```
I thought this is epoch time?

madanadit replied:
I copied the comment from UnderFIleSystem interface. Rewording this to epoch time in ms

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+  /**
+   * Information about a single object in object UFS.
+   */
+  protected class ObjectStatus {
```
Should there be a relationship between this and `UnderFileStatus`?

madanadit replied:
The information in ObjectStatus is fetched as part of listStatus in some cases. It seems like for swift, StoredObject.getContentLength might be an extra RPC (still need to confirm). If the content length, modified time comes for free for the UFSs we care about the most, we can possibly merge ObjectStatus into UnderFileStatus. In any case, I'd like to deal with it in a separate PR.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
+      return false;
+    }
+    // Source exists and is a file, and destination does not exist
```
This comment does not seem accurate?

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+      }
+    }
+    // Delete src and everything under src
+    return deleteDirectory(src, DeleteOptions.defaults().setRecursive(true));
```
Not sure if there is any overhead, but I think we can get away with only deleting the direct children.

madanadit replied:
@calvinjia I'm not sure what you mean here. Can you please clarify?

calvinjia replied:
Oops, ignore this comment.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+    if (child.startsWith(parent)) {
+      return child.substring(parent.length());
+    }
+    throw new IOException(String.format("Invalid prefix. Parent: %s Child: %s", parent, child));
```
Add to `ExceptionMessage`?

madanadit commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,671 @@
+   */
+  protected int getListingLengthMax() {
+    return 1000;
+  }
```
A static var would be the same for all understores.

calvinjia replied:
oh right, sorry

madanadit commented:
@calvinjia Please take another look.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12482/
Test PASSed.

gpang commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -0,0 +1,148 @@
+  @Override
+  public UnderFileStatus[] listStatus(String path, ListOptions options) throws IOException {
+    if (!options.isRecursive()) {
+      return listStatus(path);
```
To me it is a bit strange that the more general version (takes options) of `listStatus` calls the default version of `listStatus`.

Is the intention that UFS implementors only implement the default version of `listStatus`? I am a bit confused.

calvinjia replied:
Having the `listStatus` with options handle all the logic involving options and having implementing classes only implement `listStatus` without options makes sense to me. As a side note, I would prefer to call the method which should be implemented something like `listStatusInternal`. The question is if it is possible to do this efficiently, or will we need to expose the options.

gpang commented on core/common/src/main/java/alluxio/underfs/UnderFileStatus.java:
```diff
@@ -0,0 +1,78 @@
+   * a file or directory name.
+   *
+   * @param children array of listing statuses
+   * @return array of file or directory names
```
This method intentionally returns `null` if input is `null`, right? then this should be mentioned in this return javadoc.

gpang commented on core/common/src/main/java/alluxio/underfs/UnderFileStatus.java:
```diff
@@ -0,0 +1,78 @@
+  private final String mName;
+
+  /**
+   * Create new instance for under file information.
```
NIT: Creates

gpang commented on core/common/src/main/java/alluxio/underfs/UnderFileStatus.java:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  /**
+   * Convert an array of UFS file status to a listing result where each element in the array is
```
NIT: Converts

gpang commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,675 @@
+  /** Value used to indicate nested structure. */
+  protected static final char PATH_SEPARATOR_CHAR = '/';
+
+  /** Value used to indicate nested structure. */
```
maybe mention this is the `String` representation of `{@link PATH_SEPARATOR_CHAR}` ?

gpang commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,675 @@
+    }
+
+    /**
+     * Gets the file size in bytes.
```
since we are in the object-store world, this should be object, and content length?

gpang commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,675 @@
+  }
+
+  /**
+   * A chunk of listing results.
```
When you say "chunk" does this mean that this class only represents a portion of the full listing result? If so, should we call it such that it does not seem like it is the full listing result? `ObjectListingResult` sounds like this is the entire listing result.

madanadit replied:
Yes. This class is potentially a partial listing result. Renaming.

gpang commented on core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java:
```diff
@@ -0,0 +1,675 @@
+   *
+   * @return length of each list request
+   */
+  protected int getListingLength() {
```
Hrmmm, how are you defining "listing"? I thought a listing was the full list of names, and a chunk (the earlier datastructure) was a portion of that full list.

But now, it seems like you are referring to a chunk as a listing. Is that correct? Do you think that might be confusing?

madanadit replied:
Renaming 'Listing' -> 'ListingChunk'. You are right, this is confusing.

madanadit commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -43,36 +36,135 @@
-
-  /** A map of property names to values. */
-  protected HashMap<String, String> mProperties = new HashMap<>();
+public interface UnderFileSystem {
```
yes. leaving a TODO

madanadit commented on core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:
```diff
@@ -309,42 +256,35 @@ public OutputStream create(String path) throws IOException {
-   * @param options the options for create
-   * @return A {@code OutputStream} object
+   * @param path of the directory to delete
+   * @return true if directory was found and deleted, false otherwise
```
Yes.  A non-recursive delete only succeeds on an empty directory. Updating the comment

calvinjia commented on core/common/src/main/java/alluxio/exception/ExceptionMessage.java:
```diff
@@ -112,6 +113,7 @@
   FILE_ALREADY_EXISTS("{0} already exists"),
   FILE_CREATE_IS_DIRECTORY("{0} already exists. Directories cannot be overwritten with create"),
   HDFS_FILE_NOT_FOUND("File {0} with URI {1} is not found"),
+  PARENT_CREATION_FAILED_UFS("Unable to create parent directories for path {0}"),
```
It doesn't seem like this error message is restricted to operations with the UFS? Not sure if you want to make it specific or general though.

calvinjia commented on core/server/src/main/java/alluxio/master/file/FileSystemMaster.java:
```diff
@@ -2445,20 +2468,17 @@ private void mountInternal(LockedInodePath inodePath, AlluxioURI ufsPath, boolea
-      if (defaultUfs.exists(PathUtils.concatPath(defaultUfsPath, alluxioPath.getPath()))) {
+      UnderFileSystem defaultUfs = UnderFileSystem.Factory.get(defaultUfsPath);
+      String shadowPath = PathUtils.concatPath(defaultUfsPath, alluxioPath.getPath());
+      if (defaultUfs.isFile(shadowPath) || defaultUfs.isDirectory(shadowPath)) {
```
Seems like there are still areas where `exists` would be useful, do you plan to simplify these later?

madanadit replied:
Adding back exists for completeness of this PR

calvinjia commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:
```diff
@@ -548,263 +258,143 @@ private boolean copy(String src, String dst) {
+  @Override
+  protected ObjectListingResult getObjectListing(String key, boolean recursive)
+      throws IOException {
+    String delimiter = recursive ? "" : PATH_SEPARATOR;
```
Common prefixes which have not already been marked by Alluxio will not be fetched correctly in the recursive case?

madanadit replied:
Yes, I have a test case and fix ready. I'll open a follow up PR.

madanadit replied:
Adding a TODO for now

madanadit replied:
Turns out we never use the recursive list in the UFS, that's why we don't encounter the bug. Still I'll add the fix in the next PR.

madanadit commented:
@gpang @calvinjia Thanks for another pass!. I addressed the comments. Let me know what you guys think about listStatus w/ options calling listStatus w/o options.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/12485/
Test PASSed.

calvinjia commented:
@madanadit I commented on the list status part. These sum of the changes from the ufs-inter branch look good to me. Thanks for implementing this feature!
